middleware/http/oauth2/oauth2_middleware.go

```diff
@@ -15,148 +15,439 @@ package oauth2
+	setTokenFn func(w http.ResponseWriter, r *http.Request, reqClaims map[string]string, token string, exp time.Duration)
 
-// Middleware is an oAuth2 authentication middleware.
-type Middleware struct {
 	logger logger.Logger
+	meta   OAuth2MiddlewareMetadata
 }
 
-const (
-	stateParam   = "state"
-	savedState   = "auth-state"
-	redirectPath = "redirect-url"
-	codeParam    = "code"
-)
+// NewOAuth2Middleware returns a new OAuth2 middleware.
+func NewOAuth2Middleware(log logger.Logger) middleware.Middleware {
+	return &OAuth2Middleware{
+		logger: log,
+	}
+}
 
 // GetHandler retruns the HTTP handler provided by the middleware.
-func (m *Middleware) GetHandler(ctx context.Context, metadata middleware.Metadata) (func(next http.Handler) http.Handler, error) {
-	meta, err := m.getNativeMetadata(metadata)
+func (m *OAuth2Middleware) GetHandler(ctx context.Context, metadata middleware.Metadata) (func(next http.Handler) http.Handler, error) {
+	err := m.meta.FromMetadata(metadata, m.logger)
+	if err != nil {
+		return nil, fmt.Errorf("invalid metadata: %w", err)
+	}
+
+	// Derive the token keys
+	err = m.meta.setTokenKeys()
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to derive token keys: %w", err)
 	}
 
-	forceHTTPS := utils.IsTruthy(meta.ForceHTTPS)
-	conf := &oauth2.Config{
-		ClientID:     meta.ClientID,
-		ClientSecret: meta.ClientSecret,
-		Scopes:       strings.Split(meta.Scopes, ","),
-		RedirectURL:  meta.RedirectURL,
-		Endpoint: oauth2.Endpoint{
-			AuthURL:  meta.AuthURL,
-			TokenURL: meta.TokenURL,
-		},
+	// Create the OAuth2 configuration object
+	m.meta.setOAuth2Conf()
+
+	return m.getHandler(), nil
+}
+
+// Returns the handler. This is split out to make unit testing easier.
+func (m *OAuth2Middleware) getHandler() func(next http.Handler) http.Handler {
+	// Set the callbacks depending on the mode of operation
+	switch m.meta.Mode {
+	case modeCookie:
+		m.getTokenFn = m.GetClaimsFromCookie
+		m.setTokenFn = m.cookieModeSetTokenInResponse
+		m.claimsForAuthFn = m.cookieModeClaimsForAuth
+	case modeHeader:
+		m.getTokenFn = m.headerModeGetClaimsFromHeader
+		m.setTokenFn = m.headerModeSetTokenResponse
 	}
 
-	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			session := sessions.Start(w, r)
+	return m.handler
+}
 
-			if session.GetString(meta.AuthHeaderName) != "" {
-				r.Header.Add(meta.AuthHeaderName, session.GetString(meta.AuthHeaderName))
-				next.ServeHTTP(w, r)
+func (m *OAuth2Middleware) handler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// If we have the "state" and "code" parameter, we need to exchange the authorization code for the access token
+		code := r.URL.Query().Get("code")
+		state := r.URL.Query().Get("state")
+		if code != "" && state != "" {
+			// Always get the claims from the cookies in this case
+			claims, err := m.GetClaimsFromCookie(r)
```

Comment:
Trying to understand: Why claims are always fetched from cookie and not from header, if code and state are provided?